In [1]:
!pip install pandas

     |████████████████████████████████| 12.4 MB 24.1 MB/s eta 0:00:01
     |████████████████████████████████| 17.3 MB 46.8 MB/s eta 0:00:01
     |████████████████████████████████| 346 kB 59.8 MB/s eta 0:00:01
     |████████████████████████████████| 502 kB 50.4 MB/s eta 0:00:01
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [2]:
import pandas as pd

In [3]:
pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [4]:
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')

In [5]:
type(df)

pandas.core.frame.DataFrame

In [6]:
df.head(2)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   

   Parch     Ticket     Fare Cabin Embarked  
0      0  A/5 21171   7.2500   NaN        S  
1      0   PC 17599  71.2833   C85        C

In [7]:
df.tail(4)

PassengerId  Survived  Pclass                                      Name  \
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [8]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [9]:
df.info()
#it is similiar to dtype but with some extra features like how many times a particular column has non-null values

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [10]:
df.describe()
#describe() only considers those columns which has datatype = numeric
#25% means 25 percentile which means the data available at 25% of whole data

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [11]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [12]:
df.dtypes == "object"

PassengerId    False
Survived       False
Pclass         False
Name            True
Sex             True
Age            False
SibSp          False
Parch          False
Ticket          True
Fare           False
Cabin           True
Embarked        True
dtype: bool

In [13]:
type(df.dtypes)

pandas.core.series.Series

In [14]:
df.dtypes[df.dtypes == "object"]

Name        object
Sex         object
Ticket      object
Cabin       object
Embarked    object
dtype: object

In [15]:
df.dtypes[df.dtypes == "object"].index

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [16]:
col_name = df.dtypes[df.dtypes == "object"].index

In [17]:
col_name

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [18]:
df[col_name]

Name     Sex  \
0                              Braund, Mr. Owen Harris    male   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2                               Heikkinen, Miss. Laina  female   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4                             Allen, Mr. William Henry    male   
..                                                 ...     ...   
886                              Montvila, Rev. Juozas    male   
887                       Graham, Miss. Margaret Edith  female   
888           Johnston, Miss. Catherine Helen "Carrie"  female   
889                              Behr, Mr. Karl Howell    male   
890                                Dooley, Mr. Patrick    male   

               Ticket Cabin Embarked  
0           A/5 21171   NaN        S  
1            PC 17599   C85        C  
2    STON/O2. 3101282   NaN        S  
3              113803  C123        S  
4              373450   NaN        S  
..                ...   ...      ...  
886            211536   NaN        S  
887            112053   B42        S  
888        W./C. 6607   NaN        S  
889            111369  C148        C  
890            370376   NaN        Q  

[891 rows x 5 columns]

In [19]:
#now after filtering out column with 'object' datatype we can perform describe() function
df[col_name].describe()
#from the output we can conclude that all the names are unique i.e. non are duplicate
#for ticket col, there are total 891 count but only 681 are unique which can simply means
#more than one passanger are booked with same ticket
#top simply higest no. of recall in particular col. for e.g. male has recall 577 times and '347082' is the ticket id associated with 7 people which is max.

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

In [20]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [21]:
#adding new column to pre existing dataset
df['ineuron'] = "sudh"

In [22]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked ineuron  
0        0         A/5 21171   7.2500   NaN        S    sudh  
1        0          PC 17599  71.2833   C85        C    sudh  
2        0  STON/O2. 3101282   7.9250   NaN        S    sudh  
3        0            113803  53.1000  C123        S    sudh  
4        0            373450   8.0500   NaN        S    sudh  
..     ...               ...      ...   ...      ...     ...  
886      0            211536  13.0000   NaN        S    sudh  
887      0            112053  30.0000   B42        S    sudh  
888      2        W./C. 6607  23.4500   NaN        S    sudh  
889      0            111369  30.0000  C148        C    sudh  
890      0            370376   7.7500   NaN        Q    sudh  

[891 rows x 13 columns]

In [23]:
df['Name'][0:15]

0                               Braund, Mr. Owen Harris
1     Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                Heikkinen, Miss. Laina
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                              Allen, Mr. William Henry
5                                      Moran, Mr. James
6                               McCarthy, Mr. Timothy J
7                        Palsson, Master. Gosta Leonard
8     Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
9                   Nasser, Mrs. Nicholas (Adele Achem)
10                      Sandstrom, Miss. Marguerite Rut
11                             Bonnell, Miss. Elizabeth
12                       Saundercock, Mr. William Henry
13                          Andersson, Mr. Anders Johan
14                 Vestrom, Miss. Hulda Amanda Adolfina
Name: Name, dtype: object

In [24]:
#alternate records from 0 to 15 index
df['Name'][0:15:2]
#you can use all concept of slicing here too as series is equivallent to list

0                               Braund, Mr. Owen Harris
2                                Heikkinen, Miss. Laina
4                              Allen, Mr. William Henry
6                               McCarthy, Mr. Timothy J
8     Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
10                      Sandstrom, Miss. Marguerite Rut
12                       Saundercock, Mr. William Henry
14                 Vestrom, Miss. Hulda Amanda Adolfina
Name: Name, dtype: object

In [25]:
#in reverse order
df['Name'][14::-1]

14                 Vestrom, Miss. Hulda Amanda Adolfina
13                          Andersson, Mr. Anders Johan
12                       Saundercock, Mr. William Henry
11                             Bonnell, Miss. Elizabeth
10                      Sandstrom, Miss. Marguerite Rut
9                   Nasser, Mrs. Nicholas (Adele Achem)
8     Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
7                        Palsson, Master. Gosta Leonard
6                               McCarthy, Mr. Timothy J
5                                      Moran, Mr. James
4                              Allen, Mr. William Henry
3          Futrelle, Mrs. Jacques Heath (Lily May Peel)
2                                Heikkinen, Miss. Laina
1     Cumings, Mrs. John Bradley (Florence Briggs Th...
0                               Braund, Mr. Owen Harris
Name: Name, dtype: object

In [26]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked ineuron  
0        0         A/5 21171   7.2500   NaN        S    sudh  
1        0          PC 17599  71.2833   C85        C    sudh  
2        0  STON/O2. 3101282   7.9250   NaN        S    sudh  
3        0            113803  53.1000  C123        S    sudh  
4        0            373450   8.0500   NaN        S    sudh  
..     ...               ...      ...   ...      ...     ...  
886      0            211536  13.0000   NaN        S    sudh  
887      0            112053  30.0000   B42        S    sudh  
888      2        W./C. 6607  23.4500   NaN        S    sudh  
889      0            111369  30.0000  C148        C    sudh  
890      0            370376   7.7500   NaN        Q    sudh  

[891 rows x 13 columns]

In [27]:
#finding indexes of 'Age' col with null values
ind = df[df['Age'].isnull() == True].index

In [28]:
ind

Index([  5,  17,  19,  26,  28,  29,  31,  32,  36,  42,
       ...
       832, 837, 839, 846, 849, 859, 863, 868, 878, 888],
      dtype='int64', length=177)

In [29]:
len(df[df['Age'].isnull()==True].index)

177

In [30]:
#showing all records with null values of Age
df.loc[ind]
#we will talk abou loc later

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked ineuron  
5      male  NaN      0      0      330877   8.4583   NaN        Q    sudh  
17     male  NaN      0      0      244373  13.0000   NaN        S    sudh  
19   female  NaN      0      0        2649   7.2250   NaN        C    sudh  
26     male  NaN      0      0        2631   7.2250   NaN        C    sudh  
28   female  NaN      0      0      330959   7.8792   NaN        Q    sudh  
..      ...  ...    ...    ...         ...      ...   ...      ...     ...  
859    male  NaN      0      0        2629   7.2292   NaN        C    sudh  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S    sudh  
868    male  NaN      0      0      345777   9.5000   NaN        S    sudh  
878    male  NaN      0      0      349217   7.8958   NaN        S    sudh  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S    sudh  

[177 rows x 13 columns]

In [31]:
#without using loc
df[df['Age'].isnull()]

PassengerId  Survived  Pclass                                      Name  \
5              6         0       3                          Moran, Mr. James   
17            18         1       2              Williams, Mr. Charles Eugene   
19            20         1       3                   Masselmani, Mrs. Fatima   
26            27         0       3                   Emir, Mr. Farred Chehab   
28            29         1       3             O'Dwyer, Miss. Ellen "Nellie"   
..           ...       ...     ...                                       ...   
859          860         0       3                          Razi, Mr. Raihed   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
868          869         0       3               van Melkebeke, Mr. Philemon   
878          879         0       3                        Laleff, Mr. Kristo   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex  Age  SibSp  Parch      Ticket     Fare Cabin Embarked ineuron  
5      male  NaN      0      0      330877   8.4583   NaN        Q    sudh  
17     male  NaN      0      0      244373  13.0000   NaN        S    sudh  
19   female  NaN      0      0        2649   7.2250   NaN        C    sudh  
26     male  NaN      0      0        2631   7.2250   NaN        C    sudh  
28   female  NaN      0      0      330959   7.8792   NaN        Q    sudh  
..      ...  ...    ...    ...         ...      ...   ...      ...     ...  
859    male  NaN      0      0        2629   7.2292   NaN        C    sudh  
863  female  NaN      8      2    CA. 2343  69.5500   NaN        S    sudh  
868    male  NaN      0      0      345777   9.5000   NaN        S    sudh  
878    male  NaN      0      0      349217   7.8958   NaN        S    sudh  
888  female  NaN      1      2  W./C. 6607  23.4500   NaN        S    sudh  

[177 rows x 13 columns]

In [32]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked ineuron  
0        0         A/5 21171   7.2500   NaN        S    sudh  
1        0          PC 17599  71.2833   C85        C    sudh  
2        0  STON/O2. 3101282   7.9250   NaN        S    sudh  
3        0            113803  53.1000  C123        S    sudh  
4        0            373450   8.0500   NaN        S    sudh  
..     ...               ...      ...   ...      ...     ...  
886      0            211536  13.0000   NaN        S    sudh  
887      0            112053  30.0000   B42        S    sudh  
888      2        W./C. 6607  23.4500   NaN        S    sudh  
889      0            111369  30.0000  C148        C    sudh  
890      0            370376   7.7500   NaN        Q    sudh  

[891 rows x 13 columns]

In [33]:
#to find out max from 'Fare' col
max(df['Fare'])

512.3292

In [34]:
#getting details of persons with max fare
df[df['Fare'] == max(df['Fare'])]

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1              Lesurer, Mr. Gustave J   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  \
258  female  35.0      0      0  PC 17755  512.3292          NaN        C   
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C   
737    male  35.0      0      0  PC 17755  512.3292         B101        C   

    ineuron  
258    sudh  
679    sudh  
737    sudh

In [35]:
#getting only names of persons with max fare
df[df['Fare'] == max(df['Fare'])]['Name']

258                      Ward, Miss. Anna
679    Cardeza, Mr. Thomas Drake Martinez
737                Lesurer, Mr. Gustave J
Name: Name, dtype: object

In [36]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked ineuron  
0      0         A/5 21171   7.2500   NaN        S    sudh  
1      0          PC 17599  71.2833   C85        C    sudh  
2      0  STON/O2. 3101282   7.9250   NaN        S    sudh  
3      0            113803  53.1000  C123        S    sudh  
4      0            373450   8.0500   NaN        S    sudh

1 . Find out how many male and female passenger was onboarded .<br>
2. how many survived we have .<br>
3. how many casuality we have <br>
4 . what is name of a person who is the eldest one . <br>
5 . how many passenger do we have in first , second and third class <br>
6. how many person we have whose name starts with "s"<br>
7. try to create a new column which is a summation  of "SibSp" and "parch"<br>
8 . how many person do we have below age of 25 .<br>
9 . how many person died whose age was less then 40 <br>
10 . from a  cabin column seperate text and numeric value . <br>


In [37]:
#Ans1. Find out how many male and female passenger was onboarded.
df["Sex"].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [38]:
#Alternative

#for male
print(len(df[df["Sex"] == "male"]))

#for female
print(len(df[df["Sex"] == "female"]))

577
314


In [39]:
#Alternative2
#df.groupby("Sex").count()
df.groupby("Sex")['Sex'].count()

Sex
female    314
male      577
Name: Sex, dtype: int64

In [40]:
#Ans 2. how many survived we have 
#by Nirbhay
sum(df["Survived"])

342

In [41]:
#Alternative
df["Survived"].value_counts()

Survived
0    549
1    342
Name: count, dtype: int64

In [42]:
#Alternative2
#df[df["Survived"] == 1]
len(df[df["Survived"] == 1])

342

In [43]:
#Ans3. how many casuality we have
#df[df["Survived"] == 0]
len(df[df["Survived"] == 0])

549

In [44]:
#Ans4. what is name of a person who is the eldest one.
#df[df['Age'] == df["Age"].max()]
df[df['Age'] == df["Age"].max()]['Name']

630    Barkworth, Mr. Algernon Henry Wilson
Name: Name, dtype: object

In [45]:
#Ans5. how many passenger do we have in first, second and third class
df.groupby("Pclass")['Pclass'].count()

Pclass
1    216
2    184
3    491
Name: Pclass, dtype: int64

In [46]:
#alternative
df['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [47]:
#Ans6. how many person we have whose name starts with "s"
#with str we are converting series into string
df[df['Name'].str.startswith("S")]['Name'].count()

#or
#len(df[df['Name'].str.startswith("S")]['Name'])


#if we have find passenger details whose name starts with 'S' or 'A'
#df[df['Name'].str.startswith(("S","A"))]

86

In [48]:
#Ans7. try to create a new column which is a summation  of "SibSp" and "parch"
df['Family_Size'] = df['SibSp'] + df['Parch']
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked ineuron  Family_Size  
0      0         A/5 21171   7.2500   NaN        S    sudh            1  
1      0          PC 17599  71.2833   C85        C    sudh            1  
2      0  STON/O2. 3101282   7.9250   NaN        S    sudh            0  
3      0            113803  53.1000  C123        S    sudh            1  
4      0            373450   8.0500   NaN        S    sudh            0

In [49]:
#Ans8. how many person do we have below age of 25 .
len(df[df['Age'] < 25])

278

In [50]:
#Ans9. how many person died whose age was less then 40
len(df[(df['Survived'] == 0) & (df['Age'] < 40)])

#'and' is a logical operator for combining conditions, and '&' is a bitwise operator for performing bitwise AND operations
# on integers or element-wise logical AND on boolean DataFrames in certain contexts like pandas.

322

**Read whole text - Important**

The brackets are important in the expression '(df['column1'] > 0) & (df['column2'] < 10)' when used in the context of pandas DataFrames. They are used to explicitly define the order of operations, making it clear that the two conditions should be evaluated separately before applying the bitwise AND (&) operation.

Without the brackets, the expression might be ambiguous, and it may not be clear to the interpreter how the conditions should be grouped. The use of brackets ensures that the conditions are evaluated independently, and then the bitwise AND is applied to the resulting boolean arrays.

Here's an example to illustrate the importance of brackets:
```python
    #Without brackets
    df_result = df['column1'] > 0 & df['column2'] < 10
```

In this case, due to the precedence of the bitwise AND operator (&), the expression would be interpreted as (df['column1'] > (0 & df['column2'])) < 10, which is likely not what you intend.

To avoid potential issues and ensure the correct logic, it's recommended to use brackets to explicitly define the order of operations, as shown in the original expression.

In [51]:
#Ans10. from a  cabin column seperate text and numeric value.
#must read below markdown if you are facing any issue to understand this code

# Use str.extract to create new columns for text and numeric parts
df[['Cabin_Letter', 'Cabin_Number']] = df['Cabin'].str.extract(r'([A-Za-z]+)(\d+)', expand=True)


# Print the DataFrame with the new columns
print(df[['Cabin', 'Cabin_Letter', 'Cabin_Number']])

    Cabin Cabin_Letter Cabin_Number
0     NaN          NaN          NaN
1     C85            C           85
2     NaN          NaN          NaN
3    C123            C          123
4     NaN          NaN          NaN
..    ...          ...          ...
886   NaN          NaN          NaN
887   B42            B           42
888   NaN          NaN          NaN
889  C148            C          148
890   NaN          NaN          NaN

[891 rows x 3 columns]


The expression ([A-Za-z]+)(\d+) is a regular expression pattern used for string matching. It consists of two main parts, each enclosed in parentheses:

([A-Za-z]+): This part is capturing one or more uppercase or lowercase letters. The square brackets [A-Za-z] define a character class that includes all uppercase and lowercase letters. The + means "one or more occurrences." The entire expression within the parentheses is a capturing group, meaning it will extract the matched letters.

(\d+): This part is capturing one or more digits. The \d represents any digit (equivalent to [0-9]), and the + means "one or more occurrences." Again, the entire expression within the parentheses is a capturing group, extracting the matched digits.

So, when applied to a string, such as 'C123', the regular expression will capture 'C' as the first group and '123' as the second group. This is useful when you want to extract specific patterns (like letters and numbers) from strings.

In the code above, this regular expression is used with the 'str.extract' function in pandas to separate the 'Cabin' column into two new columns: 'Cabin_Letter' and 'Cabin_Number'. 'Cabin_Letter' will contain the extracted letters, and 'Cabin_Number' will contain the extracted numbers.


**why exapand = True?**<br>
In the context of the 'str.extract' method in pandas, the 'expand=True' parameter determines whether the result of the extraction should be returned as a DataFrame with multiple columns or as a Series with a MultiIndex.

When expand=True:

If the regular expression has one capturing group (one set of parentheses), the result is a DataFrame with one column.
If the regular expression has multiple capturing groups, the result is a DataFrame with multiple columns.
In the code I provided:
```python
    df[['Cabin_Letter', 'Cabin_Number']] = df['Cabin'].str.extract(r'([A-Za-z]+)(\d+)', expand=True)
```

Here, with expand=True, the result of the extraction will be a DataFrame with two columns: 'Cabin_Letter' and 'Cabin_Number'.

If you set 'expand=False' or omit the 'expand' parameter, the result would be a DataFrame with a single column (if there is one capturing group) or a Series (if there are multiple capturing groups), and you would need to use the str.extract method for each capturing group separately.
```python
    #Example with expand=False
    df['Cabin_Letter'], df['Cabin_Number'] = df['Cabin'].str.extract(r'([A-Za-z]+)(\d+)')
```
In summary, expand=True is used when you want the result of the extraction to be returned as a DataFrame with separate columns for each capturing group.



In [52]:
#Alternative

# Use str.extract to create new columns for text and numeric parts
df['Cabin_Letter'] = df['Cabin'].str.extract(r'([A-Za-z]+)')
df['Cabin_Number'] = df['Cabin'].str.extract(r'(\d+)')


# Print the DataFrame with the new columns
print(df[['Cabin', 'Cabin_Letter', 'Cabin_Number']])

    Cabin Cabin_Letter Cabin_Number
0     NaN          NaN          NaN
1     C85            C           85
2     NaN          NaN          NaN
3    C123            C          123
4     NaN          NaN          NaN
..    ...          ...          ...
886   NaN          NaN          NaN
887   B42            B           42
888   NaN          NaN          NaN
889  C148            C          148
890   NaN          NaN          NaN

[891 rows x 3 columns]


dataset link - https://archive.ics.uci.edu/ml/datasets/Bank+Marketing

pd.read_csv('SampleCSV/bank.csv' , sep=";")

In [53]:
df1 = pd.read_csv('SampleCSV/bank.csv' , sep=";")
df1

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

1 . how many campaign available in this dataset .<br> 
2 . how many users do we have with housing and personal loan . <br>
3. how many person do we have whose age is 60+ . <br>
4 . in which month we have trageted most of the customer . <br>
5 . which mode of call is giving you more result<br>
6 . how many enterpeures do we have in this list  <br>
7 . how many customers do we have with negative balance <br>
8. prepare a group of data based on education level . <br>


https://codeshare.io/X8PY8Y



In [54]:
#Ans1. how many campaign available in this dataset?
# here we have to tell the diff kinds of compaign available

#df1['campaign'].unique()
len(df1['campaign'].unique())



32

In [55]:
#alternative
len(set(list(df1['campaign'])))

32

In [56]:
#Ans2. how many users do we have with housing and personal loan?
len(df1[(df1['housing'] == 'yes') & (df1['loan'] == 'yes')])

406

In [57]:
#Ans3. how many person do we have whose age is 60+ ?
len(df1[df1['age'] > 60])

127

In [58]:
#Ans4. in which month we have trageted most of the customer ?

#df1['month'].value_counts()
df1['month'].value_counts().idxmax()

'may'

In [59]:
#Alternative

#df1['month'].value_counts()
#As this is automatically arranging month in desc order
#so first index will be the having highest occurences
df1['month'].value_counts().index[0]

'may'

In [60]:
#Ans5. which mode of call is giving you more result?
#df1.groupby(['contact','y'])['y'].count()
df_groupby = df1.groupby(['contact','y']).count()
df_groupby

age   job  marital  education  default  balance  housing  \
contact   y                                                                
cellular  no   2480  2480     2480       2480     2480     2480     2480   
          yes   416   416      416        416      416      416      416   
telephone no    257   257      257        257      257      257      257   
          yes    44    44       44         44       44       44       44   
unknown   no   1263  1263     1263       1263     1263     1263     1263   
          yes    61    61       61         61       61       61       61   

               loan   day  month  duration  campaign  pdays  previous  \
contact   y                                                             
cellular  no   2480  2480   2480      2480      2480   2480      2480   
          yes   416   416    416       416       416    416       416   
telephone no    257   257    257       257       257    257       257   
          yes    44    44     44        44        44     44        44   
unknown   no   1263  1263   1263      1263      1263   1263      1263   
          yes    61    61     61        61        61     61        61   

               poutcome  
contact   y              
cellular  no       2480  
          yes       416  
telephone no        257  
          yes        44  
unknown   no       1263  
          yes        61

In [61]:
df_groupby.loc['cellular']


age   job  marital  education  default  balance  housing  loan   day  \
y                                                                            
no   2480  2480     2480       2480     2480     2480     2480  2480  2480   
yes   416   416      416        416      416      416      416   416   416   

     month  duration  campaign  pdays  previous  poutcome  
y                                                          
no    2480      2480      2480   2480      2480      2480  
yes    416       416       416    416       416       416

In [62]:
#but we are interested in 'yes' only
print(df_groupby.loc['cellular','yes'][0]) #remove [0] and run to understand why we have added [0]
print(df_groupby.loc['telephone','yes'][0])
print(df_groupby.loc['unknown','yes'][0])
print(max(df_groupby.loc['cellular','yes'][0],df_groupby.loc['telephone','yes'][0],df_groupby.loc['unknown','yes'][0]))

416
44
61
416


In [63]:
#Alternative + very good method by chatgpt
#df1[df1['y'] == 'yes']['contact'].value_counts()
df1[df1['y'] == 'yes']['contact'].value_counts().idxmax()

'cellular'

In [69]:
#Ans6. how many enterpeures do we have in this list
#by Nirbhay
len(df1[df1['job'] == 'entrepreneur'])

168

In [74]:
#Alternative
#by Nirbhay
job = df1.groupby('job')['job'].value_counts()
job['entrepreneur']

168

In [76]:
#Ans7. how many customers do we have with negative balance?
len(df1[df1['balance'] < 0])

366

In [125]:
#Ans8. prepare a group of data based on education level .
df1.groupby('education')['education'].value_counts()

#alternative
#df1.groupby('education')['education'].count()

#Actually ques. is not asking about this, Correct ans. is written below

education
primary       678
secondary    2306
tertiary     1350
unknown       187
Name: count, dtype: int64

In [129]:
l = []
for i in df1['education'].unique():
    l.append(df1[df1['education'] == i])

for i in l:
    print(i)

      age            job   marital education default  balance housing loan  \
0      30     unemployed   married   primary      no     1787      no   no   
9      43       services   married   primary      no      -88     yes  yes   
18     25    blue-collar    single   primary      no     -221     yes   no   
26     55    blue-collar   married   primary      no      627     yes   no   
36     78        retired  divorced   primary      no      229      no   no   
...   ...            ...       ...       ...     ...      ...     ...  ...   
4480   23    blue-collar   married   primary      no     1158     yes   no   
4485   53    blue-collar   married   primary      no      238     yes   no   
4486   37    blue-collar   married   primary      no      378     yes   no   
4499   45    blue-collar  divorced   primary      no      942      no   no   
4503   60  self-employed   married   primary      no      362      no  yes   

        contact  day month  duration  campaign  pdays  previous

In [126]:
#Alternative by chatGPT

# Group data based on 'education' level
grouped_education = df1.groupby('education')

# Display the groups
for education, group_df in grouped_education:
    print(f"\nEducation Level: {education}")
    print(group_df)

#yha se ye smjh aya ki groupby basically ek dictionay bnata hai jisme key mei vo us column ke value rkhta hai jiske basis pei hme group bnana hai and value mei us particular value ke cooresponding dataframe
#i.e. yha pei key mei primary, secondary, tertiary & unknown rkha hai and value mei unke cooresponding dataframe


Education Level: primary
      age            job   marital education default  balance housing loan  \
0      30     unemployed   married   primary      no     1787      no   no   
9      43       services   married   primary      no      -88     yes  yes   
18     25    blue-collar    single   primary      no     -221     yes   no   
26     55    blue-collar   married   primary      no      627     yes   no   
36     78        retired  divorced   primary      no      229      no   no   
...   ...            ...       ...       ...     ...      ...     ...  ...   
4480   23    blue-collar   married   primary      no     1158     yes   no   
4485   53    blue-collar   married   primary      no      238     yes   no   
4486   37    blue-collar   married   primary      no      378     yes   no   
4499   45    blue-collar  divorced   primary      no      942      no   no   
4503   60  self-employed   married   primary      no      362      no  yes   

        contact  day month  duration 

In [124]:
# #if you wannt to save the data in single csv file for ques8:

# # Save each group to a single CSV file
# grouped_data = [group_df for _, group_df in grouped_education]
# concatenated_df = pd.concat(grouped_data)

# concatenated_df.to_csv('grouped_data1.csv', index=False)
# print("Data saved to 'grouped_data.csv'")

Data saved to 'grouped_data.csv'


In [131]:
# #It is required to run the below code
# !pip install openpyxl

     |████████████████████████████████| 249 kB 18.2 MB/s eta 0:00:01


In [132]:
# # Save each group to separate sheets in a single Excel file
# with pd.ExcelWriter('grouped_data.xlsx') as writer:
#     for education, group_df in grouped_education:
#         group_df.to_excel(writer, sheet_name=education, index=False)
#         print(f"Data for {education} level saved to sheet '{education}' in 'grouped_data.xlsx'")

Data for primary level saved to sheet 'primary' in 'grouped_data.xlsx'
Data for secondary level saved to sheet 'secondary' in 'grouped_data.xlsx'
Data for tertiary level saved to sheet 'tertiary' in 'grouped_data.xlsx'
Data for unknown level saved to sheet 'unknown' in 'grouped_data.xlsx'


In [125]:
df1

age            job  marital  education default  balance housing loan  \
0      30     unemployed  married    primary      no     1787      no   no   
1      33       services  married  secondary      no     4789     yes  yes   
2      35     management   single   tertiary      no     1350     yes   no   
3      30     management  married   tertiary      no     1476     yes  yes   
4      59    blue-collar  married  secondary      no        0     yes   no   
...   ...            ...      ...        ...     ...      ...     ...  ...   
4516   33       services  married  secondary      no     -333     yes   no   
4517   57  self-employed  married   tertiary     yes    -3313     yes  yes   
4518   57     technician  married  secondary      no      295      no   no   
4519   28    blue-collar  married  secondary      no     1137      no   no   
4520   44   entrepreneur   single   tertiary      no     1136     yes  yes   

       contact  day month  duration  campaign  pdays  previous poutcome   y  
0     cellular   19   oct        79         1     -1         0  unknown  no  
1     cellular   11   may       220         1    339         4  failure  no  
2     cellular   16   apr       185         1    330         1  failure  no  
3      unknown    3   jun       199         4     -1         0  unknown  no  
4      unknown    5   may       226         1     -1         0  unknown  no  
...        ...  ...   ...       ...       ...    ...       ...      ...  ..  
4516  cellular   30   jul       329         5     -1         0  unknown  no  
4517   unknown    9   may       153         1     -1         0  unknown  no  
4518  cellular   19   aug       151        11     -1         0  unknown  no  
4519  cellular    6   feb       129         4    211         3    other  no  
4520  cellular    3   apr       345         2    249         7    other  no  

[4521 rows x 17 columns]

Challenge Question:

Read all the resume form ineuron git in pdf, word, and try to create a dataframe with resume name as a index value, and in column i am expecting email id, linked in id, git id, skills as a result.

Note: you may have to import re (regular expression) and if you face any problem while performing this task then you can also visit regex official documentation of python for more clarity

Github link: https://github.com/iNeuronai/same-resume-year-wise